org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/IndexAwareNameEnvironment.xtend

```diff
@@ -12,30 +12,41 @@ import org.eclipse.xtext.common.types.descriptions.EObjectDescriptionBasedStubGe
 import org.eclipse.xtext.naming.QualifiedName
 import org.eclipse.xtext.resource.IResourceDescriptions
 import org.eclipse.emf.ecore.resource.Resource
+import org.eclipse.xtext.naming.IQualifiedNameConverter
 
 @FinalFieldsConstructor class IndexAwareNameEnvironment implements INameEnvironment {
 
     val Resource resource
 	val ClassLoader classLoader
 	val IResourceDescriptions resourceDescriptions
 	val EObjectDescriptionBasedStubGenerator stubGenerator
+	val ClassFileCache classFileCache
+	val IQualifiedNameConverter qualifiedNameConverter
 
-	Map<QualifiedName, NameEnvironmentAnswer> cache = newHashMap()
+	Map<String, NameEnvironmentAnswer> cache = newHashMap()
     
 	override cleanup() {
 		cache.clear
 	}
 
 	override findType(char[][] compoundTypeName) {
 		val className = QualifiedName.create(compoundTypeName.map[String.valueOf(it)])
-		return findType(className)
+		return findType(className.toString)
```

Comment:
No need to convert it to a QN in the first place.

Reply:
how to easily do the join then?

Reply:
list.join(".") ?

Reply:
i will add the concurrent hashmap to the other one

Reply:
org.eclipse.jdt.core.compiler.CharOperation.toString(char[][])

Reply:
we should decide to go for one of the two appoaches